platform/linux-generic/odp_hash_crc_gen.c

```diff
@@ -0,0 +1,222 @@
+	u8[2] = reflect_u8(u32 & 0xff);
+
+	return (u8[2] << 16) | (u8[1] << 8) | u8[0];
+}
+
+/* Reflect 16 bits */
+static inline uint32_t reflect_u16(uint32_t u32)
+{
+	uint8_t u8[4];
+
+	u8[0] = reflect_u8((u32 & 0xff00) >> 8);
+	u8[1] = reflect_u8(u32 & 0xff);
+
+	return (u8[1] << 8) | u8[0];
+}
+
+/* Generate table for a 32/24/16 bit CRCs.
+ *
+ * Based on an example in RFC 1952.
+ */
+static inline void crc_table_gen(uint32_t poly, int reflect, int width)
+{
+	uint32_t i, crc, bit, shift, msb, mask;
+
+	crc_table.width   = width;
+	crc_table.poly    = poly;
+	crc_table.reflect = reflect;
+
+	shift = width - 8;
+	mask  = 0xffffffff >> (32 - width);
+	msb   = 0x1 << (width - 1);
+
+	if (reflect) {
+		if (width == 32)
+			poly = reflect_u32(poly);
+		else if (width == 24)
+			poly = reflect_u24(poly);
+		else
+			poly = reflect_u16(poly);
+	}
+
+	for (i = 0; i < 256; i++) {
+		if (reflect) {
+			crc = i;
+
+			for (bit = 0; bit < 8; bit++) {
+				if (crc & 0x1)
+					crc = poly ^ (crc >> 1);
+				else
+					crc = crc >> 1;
+			}
+		} else {
+			crc = i << shift;
+
+			for (bit = 0; bit < 8; bit++) {
+				if (crc & msb)
+					crc = poly ^ (crc << 1);
+				else
+					crc = crc << 1;
+			}
+		}
+
+		crc_table.crc[i] = crc & mask;
+	}
+}
+
+static inline uint32_t crc_calc(const uint8_t *data, uint32_t data_len,
+				uint32_t init_val, int reflect, int width)
+{
+	uint32_t i, crc, shift;
+	uint8_t byte;
+	uint32_t mask;
+
+	shift = width - 8;
+	mask  = 0xffffffff >> (32 - width);
+
+	crc = init_val;
+
+	for (i = 0; i < data_len; i++) {
+		byte = data[i];
+
+		if (reflect) {
+			crc = crc_table.crc[(crc ^ byte) & 0xff] ^ (crc >> 8);
+		} else {
+			crc = crc_table.crc[(crc >> shift) ^ byte] ^ (crc << 8);
+			crc = crc & mask;
+		}
+	}
+
+	return crc;
+}
+
+int odp_hash_crc_gen64(const void *data_ptr, uint32_t data_len,
+		       uint64_t init_val, odp_hash_crc_param_t *crc_param,
+		       uint64_t *crc_out)
+{
+	uint32_t crc;
+	int update_table;
+	uint32_t poly = crc_param->poly;
+	uint32_t width = crc_param->width;
+	int reflect = crc_param->reflect_in;
+
+	if (odp_unlikely(crc_param->reflect_in != crc_param->reflect_out)) {
+		ODP_ERR("Odd reflection setting not supported.\n");
+		return -1;
+	}
+
+	if (odp_unlikely(width != 32 && width != 24 && width != 16)) {
+		ODP_ERR("CRC width %" PRIu32 " bits not supported.\n", width);
+		return -1;
+	}
+
+	/* TODO: fix implementation of 24 bit CRC with reflection */
+	if (odp_unlikely(width == 24 && reflect)) {
+		ODP_ERR("24 bit CRC with reflection not supported.\n");
+		return -1;
+	}
+
+	odp_rwlock_read_lock(&crc_table.rwlock);
+
+	update_table = (crc_table.width != width) ||
+		       (crc_table.poly != poly) ||
+		       (crc_table.reflect != reflect);
+
+	/* Generate CRC table if not yet generated. */
+	if (odp_unlikely(update_table)) {
+		odp_rwlock_read_unlock(&crc_table.rwlock);
+		odp_rwlock_write_lock(&crc_table.rwlock);
+
+		crc_table_gen(poly, reflect, width);
+	}
```

Comment:
This adds internal state to the crc calculation.
I think it might be better to:
 - pre-generate several typically used tables
 - use thread local storage for non-typical tables.

Reply:
There's a large number of combinations, but a single application uses only one or few. I think it's better to "harden" those ones that are commonly used with API calls (like crc32 and crc32c currently), that enables better optimization for those. This generic function is there for odd cases, and apparently has not been needed frequently as it was not implemented for a long time.

Couple of parallel tables could be added to ease the pressure on the single table. But I'd do that on top of this set (this patch set from 0 to 1 table, later from 1 to N tables).